File: 24.两两交换链表中的节点.ts
/*
 * @lc app=leetcode.cn id=24 lang=typescript
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
    const dummyHead = new ListNode(0, head);
    let node = dummyHead;
    while (node.next && node.next.next) {
        const tmp = node.next.next.next;
        node.next.next.next = node.next;
        node.next = node.next.next;
        node.next.next.next = tmp;
        node = node.next.next;
    }
    return dummyHead.next;
};
// @lc code=end

File: 24.两两交换链表中的节点.rs
/*
 * @lc app=leetcode.cn id=24 lang=rust
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if let Some(mut node1) = head {
            match node1.next.take() {
                None => return Some(node1),
                Some(mut node2) => {
                    node2.next.insert(node1).next = Self::swap_pairs(node2.next.take());
                    return Some(node2);
                }
            }
        }
        return None;
    }
}
// @lc code=end

